identity-account-storage/src/storage/stronghold.rs

```diff
@@ -213,66 +325,82 @@ impl Drop for Stronghold {
+  vault.execute(generate_key).await?;
+
+  Ok(())
 }
 
 async fn retrieve_public_key(vault: &Vault<'_>, location: &KeyLocation) -> Result<PublicKey> {
-  let procedure: iota_stronghold::procedures::PublicKey = iota_stronghold::procedures::PublicKey {
+  let procedure: procedures::PublicKey = procedures::PublicKey {
     ty: location_key_type(location),
-    private_key: location_skey(location),
+    private_key: location.into(),
   };
+
   Ok(vault.execute(procedure).await.map(|public| public.to_vec().into())?)
 }
 
 async fn sign_ed25519(vault: &Vault<'_>, payload: Vec<u8>, location: &KeyLocation) -> Result<Signature> {
-  let procedure: Ed25519Sign = Ed25519Sign {
-    private_key: location_skey(location),
+  let procedure: procedures::Ed25519Sign = procedures::Ed25519Sign {
+    private_key: location.into(),
     msg: payload,
   };
   let signature: [u8; 64] = vault.execute(procedure).await?;
   Ok(Signature::new(signature.into()))
 }
 
-fn location_key_type(location: &KeyLocation) -> iota_stronghold::procedures::KeyType {
-  match location.method() {
-    MethodType::Ed25519VerificationKey2018 => iota_stronghold::procedures::KeyType::Ed25519,
-    MethodType::X25519KeyAgreementKey2019 => iota_stronghold::procedures::KeyType::X25519,
-  }
-}
+// Moves a key from one location to another, deleting the old one.
+async fn move_key(vault: &Vault<'_>, source: Location, target: Location) -> Result<()> {
+  let copy_record = procedures::CopyRecord {
+    source: source.clone(),
+    target,
+    hint: default_hint(),
+  };
 
-fn location_chain_state() -> Location {
-  Location::generic("$chain_state", Vec::new())
-}
+  vault.execute(copy_record).await?;
 
-fn location_state() -> Location {
-  Location::generic("$state", Vec::new())
+  let revoke_data = procedures::RevokeData {
+    location: source,
+    should_gc: true,
+  };
+
+  vault.execute(revoke_data).await?;
+
+  Ok(())
 }
 
-fn location_seed(location: &KeyLocation) -> Location {
-  Location::generic(fmt_key("$seed", location), Vec::new())
+async fn get_index(store: &Store<'_>) -> Result<BTreeSet<IotaDID>> {
+  let index: BTreeSet<IotaDID> = match store.get(INDEX_STORE_KEY).await? {
+    Some(index_vec) => BTreeSet::<IotaDID>::from_json_slice(&index_vec)?,
+    None => BTreeSet::new(),
+  };
+
+  Ok(index)
 }
 
-fn location_skey(location: &KeyLocation) -> Location {
-  Location::generic(fmt_key("$skey", location), Vec::new())
+async fn set_index(store: &Store<'_>, index: BTreeSet<IotaDID>) -> Result<()> {
+  let index_vec: Vec<u8> = index.to_json_vec()?;
+
+  store.set(INDEX_STORE_KEY, index_vec, None).await?;
+
+  Ok(())
 }
 
-fn fmt_key(prefix: &str, location: &KeyLocation) -> Vec<u8> {
-  format!("{}:{}:{}", prefix, location.generation(), location.fragment_name()).into_bytes()
+impl From<&KeyLocation> for Location {
+  fn from(key_location: &KeyLocation) -> Self {
+    Location::generic(VAULT_PATH.to_vec(), key_location.to_string().into_bytes())
+  }
 }
```

Comment:
I don't think this should use `to_string` since the `Display` implementation could be changed without anyone thinking it affects the `Stronghold` locations.

Reply:
That's fair. Each storage can derive a location on their own based on `KeyLocation::fragment` and `KeyLocation::key_hash`. In that case, I will remove the `Display` impl entirely, since there's no reason to have it then.

Reply:
It's not about removing `Display`, it's not using the implementation for logic. We can still use it in `DuplicateKeyLocation` and errors if needed. Error message strings are not intended to be stable in terms of API stability when considering breaking changes. Error variants are breaking changes, however.

Reply:
It seems like good practice to me to have a stable `Display` implementation. If we don't commit to use `Display` as the canonical representation of a location, then I'm not sure if it still serves any purpose. If any storage is expected to use a different representation internally, a debug message containing our custom `Display` implementation doesn't help, `Debug` is better, even.

Reply:
`Display` should be as stable as the struct itself. As its namesake indicates, it should only be used for `Display` purposes, i.e. human-readable string representations. `Debug` string representations are generally considered not stable in `Rust`, even for standard library objects.

As previously suggested, we can offer a similar function for a canonical string representation suitable for `Storage` locations on `KeyLocation` that does not depend on `Display`.